src/hardware/input/intel8042.cpp

```diff
@@ -0,0 +1,1037 @@
+	bit_view<4, 1> is_disabled_kbd;
+	// bit 5: 1 = aux (mouse) port disabled
+	bit_view<5, 1> is_disabled_aux;
+	// bit 6: 1 = keyboard input should be translated for XT comaptibility
+	bit_view<6, 1> uses_kbd_translation;
+	// bit 7: reserved, should be 0
+
+} config_byte;
+
+static void sanitize_config_byte()
+{
+	config_byte.data = (config_byte.data & 0b0111'0111) | 0b0000'0100;
+}
+
+static auto& is_irq_active_kbd    = config_byte.is_irq_active_kbd;
+static auto& is_irq_active_aux    = config_byte.is_irq_active_aux;
+static auto& is_disabled_kbd      = config_byte.is_disabled_kbd;
+static auto& is_disabled_aux      = config_byte.is_disabled_aux;
+static auto& uses_kbd_translation = config_byte.uses_kbd_translation;
+
+// Byte returned from port 0x60
+
+static uint8_t data_byte = 0;
+
+// Byte returned from port 0x64
+
+static union {
+	uint8_t data = 0b0001'1100;
+
+	// bit 0: 1 = new byte is waiting in 0x60
+	bit_view<0, 1> is_data_new;
+	// bit 1: input buffer status, 0 = guest can write to 0x60 or 0x64
+	// bit 2: 1 = POST has already passed since power on
+	// bit 3: 0 = last write was to 0x60, 1 = to 0x64
+	bit_view<3, 1> was_last_write_cmd;
+
+	// bit 4: 0 = keyboard locked (by external switch)
+	// bit 5: 1 = data byte in 0x60 is from AUX
+	bit_view<5, 1> is_data_from_aux;
+	// bit 6: 1 = timeout error during data transmission
+	bit_view<6, 1> is_transmit_timeout;
+	// bit 7: 1 = parity error
+
+} status_byte;
+
+static auto& is_data_new      = status_byte.is_data_new;
+static auto& is_data_from_aux = status_byte.is_data_from_aux;
+static bool is_data_from_kbd  = false; // not present in status byte
+
+static auto& is_transmit_timeout = status_byte.is_transmit_timeout;
+
+// Controller internal buffer
+
+static struct {
+	uint8_t data = 0;
+
+	bool is_from_aux = false;
+	bool is_from_kbd = false;
+	bool skip_delay  = false;
+
+} buffer[buffer_size];
+
+static size_t buffer_start_idx = 0;
+static size_t buffer_num_used  = 0;
+
+static size_t waiting_bytes_from_aux = 0;
+static size_t waiting_bytes_from_kbd = 0;
+
+// true = delay timer is in progress
+static bool delay_running = false;
+// true = delay timer expired, event can be sent immediately
+static bool delay_finished = true;
+
+// Executing command, do not notify devices about readiness for accepting frame
+static bool should_skip_device_notify = false;
+
+// Command currently being executed, waiting for parameter
+static Command current_command = Command::None;
+
+// ***************************************************************************
+// Helper routines to log various warnings
+// ***************************************************************************
+
+static void warn_buffer_full()
+{
+	static constexpr uint32_t threshold = 15 * 1000; // 15 seconds
+
+	static bool already_warned = false;
+	static uint32_t last_timestamp = 0;
+
+	if (!already_warned || (PIC_Ticks - last_timestamp > threshold)) {
+		LOG_INFO("I8042: Internal buffer overflow");
```

Comment:
Please use warn log level for all warning log entries in this file.

Reply:
OK, used WARN level.